crates/polars-io/src/ipc/object_store_reader.rs

```diff
@@ -0,0 +1,139 @@
+use std::sync::Arc;
+
+use bytes::Bytes;
+use object_store::path::Path as ObjectPath;
+use object_store::ObjectStore;
+use polars_core::frame::DataFrame;
+use polars_error::{to_compute_err, PolarsResult};
+
+use crate::mmap::MmapBytesReader;
+use crate::pl_async::with_concurrency_budget;
+use crate::predicates::PhysicalIoExpr;
+use crate::prelude::IpcReader;
+use crate::{RowIndex, SerReader};
+
+#[derive(Debug, Clone)]
+enum Columns {
+    Names(Vec<String>),
+    Indices(Vec<usize>),
+}
+
+#[derive(Default, Clone)]
+pub struct IpcReadOptions {
+    columns: Option<Columns>,
+    row_limit: Option<usize>,
+    row_index: Option<RowIndex>,
+    predicate: Option<Arc<dyn PhysicalIoExpr>>,
+}
+
+impl IpcReadOptions {
+    pub fn column_names(mut self, names: impl Into<Option<Vec<String>>>) -> Self {
+        self.columns = names.into().map(Columns::Names);
+        self
+    }
+
+    pub fn column_indices(mut self, indices: impl Into<Option<Vec<usize>>>) -> Self {
+        self.columns = indices.into().map(Columns::Indices);
+        self
+    }
+
+    pub fn row_limit(mut self, row_limit: impl Into<Option<usize>>) -> Self {
+        self.row_limit = row_limit.into();
+        self
+    }
+
+    pub fn row_index(mut self, row_index: impl Into<Option<RowIndex>>) -> Self {
+        self.row_index = row_index.into();
+        self
+    }
+
+    pub fn predicate(mut self, predicate: impl Into<Option<Arc<dyn PhysicalIoExpr>>>) -> Self {
+        self.predicate = predicate.into();
+        self
+    }
+}
+
+pub async fn read_ipc_async<O: ObjectStore>(
+    store: &O,
+    path: &str,
+    options: IpcReadOptions,
+) -> PolarsResult<DataFrame> {
+    let path = ObjectPath::parse(path).map_err(to_compute_err)?;
+
+    // TODO: Load only what is needed, rather than everything.
+    let file_bytes = with_concurrency_budget(1, || read_bytes(store, &path)).await?;
+    
+    read_ipc(std::io::Cursor::new(file_bytes), options)
+}
+
+async fn read_bytes<O: ObjectStore>(store: &O, path: &ObjectPath) -> PolarsResult<Bytes> {
+    // TODO: Is `to_compute_err` appropriate? It is used in the Parquet
+    // reader as well but I am not sure it is what we want.
+    let get_result = store.get(path).await.map_err(to_compute_err)?;
```

Comment:
Here we should split up the work. We should be able download concurrently and the first chunks should be downloaded first.  We should have a `batched` interface that can already get `DataFrame`s of the chunks that are already downloaded.